File: api/src/main/java/com/datastrato/gravitino/rel/expressions/literals/Literals.java
/*
 * Copyright 2024 Datastrato Pvt Ltd.
 * This software is licensed under the Apache License version 2.
 */
package com.datastrato.gravitino.rel.expressions.literals;

import com.datastrato.gravitino.rel.types.Decimal;
import com.datastrato.gravitino.rel.types.Type;
import com.datastrato.gravitino.rel.types.Types;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Objects;

/** The helper class to create literals to pass into Gravitino. */
public class Literals {

  /** Used to represent a null literal. */
  public static final Literal<Types.NullType> NULL = new LiteralImpl<>(null, Types.NullType.get());

  /**
   * Creates a literal with the given value and data type.
   *
   * @param value the literal value
   * @param dataType the data type of the literal
   * @return a new {@link com.datastrato.gravitino.rel.expressions.Literal} instance
   * @param <T> the JVM type of value held by the literal
   */
  public static <T> LiteralImpl<T> of(T value, Type dataType) {
    return new LiteralImpl<>(value, dataType);
  }

  /**
   * Creates a boolean type literal with the given value.
   *
   * @param value the boolean literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Boolean> booleanLiteral(Boolean value) {
    return of(value, Types.BooleanType.get());
  }

  /**
   * Creates a byte type literal with the given value.
   *
   * @param value the byte literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Byte> byteLiteral(Byte value) {
    return of(value, Types.ByteType.get());
  }

  /**
   * Creates a short type literal with the given value.
   *
   * @param value the short literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Short> shortLiteral(Short value) {
    return of(value, Types.ShortType.get());
  }

  /**
   * Creates an integer type literal with the given value.
   *
   * @param value the integer literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Integer> integerLiteral(Integer value) {
    return of(value, Types.IntegerType.get());
  }

  /**
   * Creates a long type literal with the given value.
   *
   * @param value the long literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Long> longLiteral(Long value) {
    return of(value, Types.LongType.get());
  }

  /**
   * Creates a float type literal with the given value.
   *
   * @param value the float literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Float> floatLiteral(Float value) {
    return of(value, Types.FloatType.get());
  }

  /**
   * Creates a double type literal with the given value.
   *
   * @param value the double literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Double> doubleLiteral(Double value) {
    return of(value, Types.DoubleType.get());
  }

  /**
   * Creates a decimal type literal with the given value.
   *
   * @param value the decimal literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<Decimal> decimalLiteral(Decimal value) {

    return of(value, Types.DecimalType.of(value.precision(), value.scale()));
  }

  /**
   * Creates a date type literal with the given value.
   *
   * @param value the date literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<LocalDate> dateLiteral(LocalDate value) {
    return of(value, Types.DateType.get());
  }

  /**
   * Creates a time type literal with the given value.
   *
   * @param value the time literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<LocalTime> timeLiteral(LocalTime value) {
    return of(value, Types.TimeType.get());
  }

  /**
   * Creates a timestamp type literal with the given value.
   *
   * @param value the timestamp literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<LocalDateTime> timestampLiteral(LocalDateTime value) {
    return of(value, Types.TimestampType.withoutTimeZone());
  }

  /**
   * Creates a timestamp type literal with the given value.
   *
   * @param value the timestamp literal value, must be in the format "yyyy-MM-ddTHH:mm:ss"
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<LocalDateTime> timestampLiteral(String value) {
    return timestampLiteral(LocalDateTime.parse(value));
  }

  /**
   * Creates a string type literal with the given value.
   *
   * @param value the string literal value
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<String> stringLiteral(String value) {
    return of(value, Types.StringType.get());
  }

  /**
   * Creates a varchar type literal with the given value.
   *
   * @param value the string literal value
   * @param length the length of the varchar
   * @return a new {@link Literal} instance
   */
  public static LiteralImpl<String> varcharLiteral(int length, String value) {
    return of(value, Types.VarCharType.of(length));
  }

  /**
   * Creates a literal with the given type value.
   *
   * @param <T> The JVM type of value held by the literal.
   */
  public static final class LiteralImpl<T> implements Literal<T> {
    private final T value;
    private final Type dataType;

    private LiteralImpl(T value, Type dataType) {
      this.value = value;
      this.dataType = dataType;
    }

    @Override
    public T value() {
      return value;
    }

    @Override
    public Type dataType() {
      return dataType;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      LiteralImpl<?> literal = (LiteralImpl) o;
      return Objects.equals(value, literal.value) && Objects.equals(dataType, literal.dataType);
    }

    @Override
    public int hashCode() {
      return Objects.hash(value, dataType);
    }

    @Override
    public String toString() {
      return "LiteralImpl{" + "value=" + value + ", dataType=" + dataType + '}';
    }
  }

  private Literals() {}
}

File: clients/client-python/gravitino/api/expressions/literals.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from abc import ABC, abstractmethod
from typing import Union
from datetime import date, time, datetime


# Abstract base class for Literal
class Literal(ABC):
    """
    Represents a constant literal value in the expression API.
    """

    @abstractmethod
    def value(self) -> Union[int, float, str, datetime, time, date, bool]:
        """
        Returns the literal value.
        """
        pass

    @abstractmethod
    def data_type(self) -> str:
        """
        Returns the data type of the literal.
        """
        pass

    def children(self):
        """
        Returns the child expressions. By default, this is an empty list.
        """
        return []


# Concrete implementation of Literal
class LiteralImpl(Literal):
    def __init__(
        self, value: Union[int, float, str, datetime, time, date, bool], data_type: str
    ):
        self._value = value
        self._data_type = data_type

    def value(self) -> Union[int, float, str, datetime, time, date, bool]:
        return self._value

    def data_type(self) -> str:
        return self._data_type

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, LiteralImpl):
            return False
        return (self._value == other._value) and (self._data_type == other._data_type)

    def __hash__(self):
        return hash((self._value, self._data_type))

    def __str__(self):
        return f"LiteralImpl(value={self._value}, data_type={self._data_type})"


# Helper class to create literals
class Literals:
    @staticmethod
    def null_literal() -> Literal:
        return LiteralImpl(None, "NullType")

    @staticmethod
    def boolean_literal(value: bool) -> Literal:
        return LiteralImpl(value, "Boolean")

    @staticmethod
    def byte_literal(value: int) -> Literal:
        return LiteralImpl(value, "Byte")

    @staticmethod
    def short_literal(value: int) -> Literal:
        return LiteralImpl(value, "Short")

    @staticmethod
    def integer_literal(value: int) -> Literal:
        return LiteralImpl(value, "Integer")

    @staticmethod
    def long_literal(value: int) -> Literal:
        return LiteralImpl(value, "Long")

    @staticmethod
    def float_literal(value: float) -> Literal:
        return LiteralImpl(value, "Float")

    @staticmethod
    def double_literal(value: float) -> Literal:
        return LiteralImpl(value, "Double")

    @staticmethod
    def decimal_literal(value: float) -> Literal:
        return LiteralImpl(value, "Decimal")

    @staticmethod
    def string_literal(value: str) -> Literal:
        return LiteralImpl(value, "String")

    @staticmethod
    def varchar_literal(length: int, value: str) -> Literal:
        return LiteralImpl(value, f"Varchar({length})")

    @staticmethod
    def date_literal(value: date) -> Literal:
        return LiteralImpl(value, "Date")

    @staticmethod
    def time_literal(value: time) -> Literal:
        return LiteralImpl(value, "Time")

    @staticmethod
    def timestamp_literal(value: datetime) -> Literal:
        return LiteralImpl(value, "Timestamp")

    @staticmethod
    def timestamp_literal_from_string(value: str) -> Literal:
        return Literals.timestamp_literal(datetime.fromisoformat(value))
